Enable expansion of ${XXX} constructs

thallgren opened the pull request:
Environment variables containing `${XXX}` can be expanded using the
standard `os.Expand(key string, func(string) string) string` function,
but at present, the `envconfig` will not fully facilitate this. If
an expansion is added to a `default=`, then the unexpanded value will
be passed to the associated parser during validation. Consequently, a
declaration like this one:
```golang
AuthURL *url.URL `env:"AUTH_URL, parser=absolute-URL, default=https://${DOMAIN}/auth"`
```
will always fail since '${}' isn't allowed in a URL address.

This commit fixes this by:
1. Disabling validation of default values that contain expansions during
   parser generation. Such values cannot be validated until used.
2. Ensuring that the ${XXX} notation can be used in defaults by
   expanding such values before passing them to the parser.

thallgren committed:
Introduce a testEnv type and replace all os.LookupEnv calls

Introduces a simple `map[string]string` type with a
`lookup(string) (string, bool)` function to make testing without using
`os.LookupEnv` easier.

LukeShu commented:
I really like this feature, but I really don't like that it can no longer separate generation of the parser from running the parser. I' going to have to sit on it for a bit.

thallgren commented:
Yes, if the defaults contains expansions, then there will be lookups in the generation step. Only way to avoid that is to move that part of the validation out of the generation.

thallgren committed:
Enable expansion of ${XXX} constructs

Environment variables containing `${XXX}` can be expanded using the
standard `os.Expand(key string, func(string) string) string` function,
but at present, the `envconfig` will not fully facilitate this. If
an expansion is added to a `default=`, then the unexpanded value will
be passed to the associated parser during validation. Consequently, a
declaration like this one:
```
AuthURL *url.URL `env:"AUTH_URL, parser=absolute-URL, default=https://${DOMAIN}/auth"`
```
will always fail since '${}' isn't allowed in a URL address.

This commit fixes this by:
1. Disabling validation of default values that contain expansions during
   parser generation. Such values cannot be validated until used.
2. Ensuring that the ${XXX} notation can be used in defaults by
   expanding such values before passing them to the parser.

thallgren commented:
@LukeShu I came up with a different solution that I think has a better separation of concern. This solution will not do lookups during parser generation. Instead, it simply skips the validation of default values that contain expanded variables. There's no way to tell what they will contain until they are actually put to use. I did contemplate validating defaults even when they are not used during the parse phase, but I think that would be dead wrong too, because if the default isn't used, then there's no requirement that the expanded variables that it contains actually exists.

I'm happier with this approach.

LukeShu commented on envconfig.go:
```diff
@@ -240,30 +256,31 @@ func generateFieldHandler(i int, tag envTag, typeHandler FieldTypeHandler) func(
 			var ev string
 			if ev, found = lookup(tag.Name); found {
-				val, err = typeHandler.Parsers[parser](ev)
+				val, err = typeHandler.Parsers[parser](expand(ev, lookup))
```
This is a little weird to me.

If I set
```bash
V1='foo'
V2='${V1}' # note the single-quotes, not double-quotes
```
It'd be very surprising to me that a program looking at `V2` would evaluate it as `foo`.

LukeShu replied:
> I don't see the recursion. `ev` origins from `lookup` which is the original environment lookup. It's expanded once on line 259, and then passed on to the parser. The resulting `val` is never expanded.

I guess calling this recursion isn't quite right, as it doesn't keep calling itself.

But in my above example, `ev` will be set to `${V1}` (which should be the terminal value), and then `expand` would return `foo` (which is wrong).

I believe this line should just be `val, err = typeHandler.Parsers[parser](ev)`.

thallgren replied:
Ah, so you want the expansion to be limited to the defaults. No expansion on values that we lookup?

thallgren replied:
As a user, I'd find that surprising, because then we treat the defaults different from the actual values.

thallgren replied:
Nah, you're right. Changing.

LukeShu commented on envconfig.go:
```diff
@@ -104,6 +105,17 @@ func (h FieldTypeHandler) parserNames() []string {
+		}
+		return ""
+	})
+}
```
Having the default here be an empty string... It's not bad, but I prickle at it. Would it be possible to do similar as `defaultFrom` to have it reference a different struct field?

thallgren replied:
I don't see the recursion. `ev` origins from `lookup` which is the original environment lookup. It's expanded once on line 259, and then passed on to the parser. The resulting `val` is never expanded.

thallgren replied:
I don't want environment expansion to reference a struct field. That would IMO violate what environment expansion is all about. It expands environment variables using the scope of the environment. The `defaultFrom` is different, because it fetches an already parsed value and uses that as the default without parsing it again, which is fine. Different scope (on the other side of the parsers).

LukeShu reviewed (changes requested):
At this point, I don't think I'm willing to approve recursively expanding variables (on line 243/259). But otherwise, LGTM.

Can you rebase up to `main` so that CI is happy?

thallgren committed:
Enable expansion of ${XXX} constructs

Environment variables containing `${XXX}` can be expanded using the
standard `os.Expand(key string, func(string) string) string` function,
but at present, the `envconfig` will not fully facilitate this. If
an expansion is added to a `default=`, then the unexpanded value will
be passed to the associated parser during validation. Consequently, a
declaration like this one:
```
AuthURL *url.URL `env:"AUTH_URL, parser=absolute-URL, default=https://${DOMAIN}/auth"`
```
will always fail since '${}' isn't allowed in a URL address.

This commit fixes this by:
1. Disabling validation of default values that contain expansions during
   parser generation. Such values cannot be validated until used.
2. Ensuring that the ${XXX} notation can be used in defaults by
   expanding such values before passing them to the parser.
